[FLINK-31660][table] add jayway json path dependency to table-planner

nateab opened the pull request:
## What is the purpose of the change

This PR adds the `json-path` dependency to the table-planner pom so that `ITCases` in the kafka connector module can run in the Intellij IDE. This is due to an old Intellij bug that does not handle the maven-shade plugin correctly https://youtrack.jetbrains.com/issue/IDEA-93855.

Original PR discussion here https://github.com/apache/flink/pull/22313

## Brief change log

  - Add the `json-path` to the `flink-table-planner_${scala.binary.version}` pom
 

## Verifying this change

This change is already covered by existing tests, such as `KafkaChangelogTableITCase`.


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

zentol commented on flink-table/flink-table-planner/pom.xml:
```diff
@@ -301,6 +301,13 @@ under the License.
+			<groupId>com.jayway.jsonpath</groupId>
+			<artifactId>json-path</artifactId>
+			<version>${jsonpath.version}</version>
+			<scope>test</scope>
```
This seems incorrect. If it has a test scope then the loader won't bundle it, nor will it ever make it's way to the kafka tests (since test dependencies aren't transitive).

nateab replied:
Okay gotcha, i've removed the test scope. However, i am no longer able to get the `KafkaChangelogTableITCase` test running in the IDE anymore, even after rebuilding the modules on the command line. I have tried adding it individually and in various combinations in `flink-table-planner-loader-bundle, flink-table-planner-loader, flink-table-planner_${scala.binary.version}`. Did you have to do anything else to get it to work?

nateab replied:
actually wait a second, i was only rebuilding the flink-table-planner module, i will try also rebuilding the other modules

nateab replied:
Nvm that wasn't it, i had to add `<include>com.jayway.jsonpath:json-path</include>` to get it to work

nateab replied:
CI failed with 
```
Dependency convergence error for com.jayway.jsonpath:json-path:jar:2.4.0:runtime paths to dependency are:
+-org.apache.flink:flink-table-planner_2.12:jar:1.17-SNAPSHOT
  +-org.apache.calcite:calcite-core:jar:1.29.0:compile
    +-com.jayway.jsonpath:json-path:jar:2.4.0:runtime
and
+-org.apache.flink:flink-table-planner_2.12:jar:1.17-SNAPSHOT
  +-com.jayway.jsonpath:json-path:jar:2.6.0:compile
```
maybe i need to try adding the dependency in the `flink-table-planner-loader` pom instead?

zentol replied:
Add an entry to the dependencyManagement of the flink-table module for json-path.

zentol replied:
_or_ use json-path 2.4.0 (because that's what calcite expects anyway). In fact, we don't need the additional dependency in the pom; the include alone would solve this.

We should use a different relocation pattern though to ensure this doesn't clash with the json-path dependency of the table runtime.

zentol commented on flink-table/flink-table-planner-loader/pom.xml:
```diff
@@ -44,13 +44,11 @@
 			<groupId>org.apache.flink</groupId>
 			<artifactId>flink-test-utils-junit</artifactId>
 		</dependency>
-
```
please revert these formatting changes

nateab replied:
reverted

nateab commented:
@flinkbot run azure

nateab commented:
CI failed with this, will look at fixing the notice file
```01:53:59,901 INFO  org.apache.flink.tools.ci.licensecheck.NoticeFileChecker     [] - Problems were detected for a NOTICE file.
	flink-table-planner:
		 These issue are legally problematic and MUST be fixed: 
			Dependency com.jayway.jsonpath:json-path:2.4.0 is not listed.```

nateab commented:
CI was failing upstream because of an outdated library, which was fixed in this commit https://github.com/apache/flink/commit/ed7ca22efc68203d4882887a3856434137e6980f, gonna rerun the CI
@flinkbot run azure

nateab commented:
@zentol I have changed the version to be 2.4.0 and updated the notice file so that CI passes, could you take another look please

zentol commented on flink-table/flink-table-planner/pom.xml:
```diff
@@ -301,6 +301,11 @@ under the License.
 			<version>${project.version}</version>
 			<scope>test</scope>
 		</dependency>
+		<dependency>
+			<groupId>com.jayway.jsonpath</groupId>
+			<artifactId>json-path</artifactId>
+			<version>2.4.0</version>
+		</dependency>
```
You now have to dependencies on json-path (table-planner/-runtime respectively), both defining the version but using the same relocation pattern, which could lead to conflicts at runtime.

You want to remove jayway from the `PlannerModule#KNOWN_MODULE_ASSOCIATIONS` so the planner doesn't have access to this dependency from the table-runtime. 

I would suggest to add a dependencyManagement to `flink-table` to control the version (and then remove the explicit dependency here and just pull it in via calcite).

nateab replied:
I have removed jayway from the `PlannerModule#KNOWN_MODULE_ASSOCIATIONS` and removed the explicit dependency in the planner, and added a dependency entry in `flink-table`

nateab commented on flink-table/flink-table-planner/pom.xml:
```diff
@@ -392,6 +397,8 @@ under the License.
 							<include>org.apache.flink:flink-sql-parser</include>
 							<include>org.apache.flink:flink-sql-parser-hive</include>
+
+							<include>com.jayway.jsonpath:json-path</include>
```
done, moved things around

nateab commented:
@zentol i believe i've addressed all your feedback, and I got CI to be pass again, so if you could take a final look hopefully 🤞 And thanks for all your help reviewing!

zentol commented:
Oh wow it did actually merge it before but failed to close the PR :facepalm:
